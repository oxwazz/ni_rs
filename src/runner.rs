use std::path::PathBuf;

#[derive(Default)]
pub(crate) struct RunnerContext {
    pub(crate) programmatic: Option<bool>,
    pub(crate) has_lock: Option<bool>,
    pub(crate) cwd: Option<PathBuf>,
}
