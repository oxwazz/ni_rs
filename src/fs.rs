use crate::runner::RunnerContext;
use serde::Deserialize;
use serde_json::Value;
use std::fs::File;
use std::io::Write;
use std::path::PathBuf;
use std::{env, process};

#[derive(Debug, Deserialize)]
struct PackageJson {
    scripts: Option<Value>,
    #[serde(rename = "scripts-info")]
    scripts_info: Option<Value>,
    dependencies: Option<Value>,
    #[serde(rename = "devDependencies")]
    dev_dependencies: Option<Value>,
}

fn get_package_json(ctx: Option<RunnerContext>) -> Option<PackageJson> {
    // Get user custom cwd or using current directory
    // example: /home/oxwazz/ni_rs/
    let cwd = ctx
        .unwrap_or_default()
        .cwd
        .unwrap_or_else(|| env::current_dir().unwrap_or(PathBuf::from(".")));

    // path = user directory + package.json
    // example: /home/oxwazz/ni_rs/package.json
    let path = cwd.join("package.json");

    // Early return error
    if !path.is_file() {
        eprintln!("Cannot find package.json");
        process::exit(1)
    }

    // Read package.json file
    let raw = File::open(path).unwrap_or_else(|_| {
        eprintln!("Failed to read package.json");
        process::exit(1)
    });

    // Parse package.json value to `PackageJson` struct
    let data: Option<PackageJson> = serde_json::from_reader(raw).unwrap_or_else(|_| {
        eprintln!("Failed to parse package.json");
        process::exit(1)
    });
    // return
    data
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_package_json_works() {
        let result = get_package_json(Some(RunnerContext {
            cwd: Some(PathBuf::from("./src/test").canonicalize().unwrap()),
            programmatic: None,
            has_lock: None,
        }));
        insta::assert_debug_snapshot!(result);
    }
}
