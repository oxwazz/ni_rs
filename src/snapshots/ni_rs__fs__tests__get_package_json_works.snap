---
source: src/fs.rs
assertion_line: 63
expression: result
snapshot_kind: text
---
Some(
    PackageJson {
        scripts: Some(
            Object {
                "build": String("unbuild"),
                "dev": String("tsx src/commands/ni.ts"),
                "lint": String("eslint .."),
                "nr": String("tsx src/commands/nr.ts"),
                "prepare": String("npx simple-git-hooks"),
                "prepublishOnly": String("npm run build"),
                "release": String("bumpp && npm publish"),
                "stub": String("unbuild --stub"),
                "test": String("vitest"),
                "typecheck": String("tsc"),
            },
        ),
        scripts_info: None,
        dependencies: None,
        dev_dependencies: Some(
            Object {
                "@antfu/eslint-config": String("^3.9.1"),
                "@posva/prompts": String("^2.4.4"),
                "@types/fs-extra": String("^11.0.4"),
                "@types/ini": String("^4.1.1"),
                "@types/node": String("^22.9.0"),
                "@types/which": String("^3.0.4"),
                "bumpp": String("^9.8.1"),
                "eslint": String("^9.15.0"),
                "fs-extra": String("^11.2.0"),
                "fzf": String("^0.5.2"),
                "ini": String("^5.0.0"),
                "lint-staged": String("^15.2.10"),
                "package-manager-detector": String("^0.2.4"),
                "picocolors": String("^1.1.1"),
                "simple-git-hooks": String("^2.11.1"),
                "taze": String("^0.18.0"),
                "terminal-link": String("^3.0.0"),
                "tinyexec": String("^0.3.1"),
                "tinyglobby": String("^0.2.10"),
                "tsx": String("^4.19.2"),
                "typescript": String("^5.6.3"),
                "unbuild": String("^2.0.0"),
                "vitest": String("^2.1.5"),
                "which": String("^5.0.0"),
            },
        ),
    },
)
